screenpipe-actions/src/main.rs

```diff
@@ -0,0 +1,241 @@
+mod screenshot;
+
+use monitor_keystroke_commands::{run_keystroke_monitor, KeystrokeCommand};
+use tokio::sync::mpsc;
+use tokio::process::Command;
+use call_ai::{call_ai_with_screenshot, AIProvider};
+use reqwest;
+use serde_json::Value;
+use anyhow::Context;
+use type_and_animate::{type_slowly, delete_characters};
+use crate::screenshot::capture_main_window_screenshot;
+use base64::{Engine as _, engine::general_purpose};
+use image::DynamicImage;
+
+async fn run_swift_script() -> anyhow::Result<String> {
+    let output = Command::new("swift")
+        .arg("src/print_all_attributes.swift")
+        .output()
+        .await?;
+
+    if !output.status.success() {
+        anyhow::bail!("swift script failed: {}", String::from_utf8_lossy(&output.stderr));
+    }
+
+    let stdout = String::from_utf8_lossy(&output.stdout).to_string();
+    println!("swift script output: {}", stdout);
+    Ok(stdout)
+}
+
+async fn search_localhost(query: &str) -> anyhow::Result<String> {
+    let client = reqwest::Client::new();
+    let url = format!("http://localhost:3030/search?q={}&content_type=all&limit=5&offset=0", query);
+    let response = client.get(&url).send().await?.text().await?;
+    Ok(response)
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    println!("starting keystroke monitor. press '//' to print attributes for the current app and call openai.");
+    let (tx, mut rx) = mpsc::channel(100);
+    
+    tokio::spawn(async move {
+        if let Err(error) = run_keystroke_monitor(tx).await {
+            eprintln!("error in keystroke monitoring: {:?}", error);
+        }
+    });
+
+    while let Some(command) = rx.recv().await {
+        match command {
+            KeystrokeCommand::DoubleSlash => {
+                println!("double slash detected. capturing screenshot and calling ai...");
+
+                // Spawn a task for run_swift_script
+                let swift_task = tokio::spawn(run_swift_script());
+
+                let screenshot = capture_main_window_screenshot()?;
+                type_slowly("thinking").await?;
+
+                let prompt = r#"Based on the desktop screenshot, provide a concise response.
+                Output your response in JSON format as follows:
+                {
+                    "response": "Your response text here"
+                }
+                The response should be casual, social media style.
+                Do no be salesy.
+                Be emotional, capture emotions of the context.
+                The response should match the length of the message we are responding to.
+                It should not look like it was written by AI.
+                "#.to_string();
+
+                // Choose the AI provider here
+                let provider = AIProvider::OpenAI; // or AIProvider::Claude
+
+                // Wrap the AI call in a Future
+                let ai_task = tokio::spawn(call_ai_with_screenshot(provider, prompt, true, screenshot));
```

Comment:
improvement: dont use json output, just stream text, pass a channel to the function for example and then stream the text to the typing